### DISCUSSION TITLE
Change import in assets/furo.js to be usable as-is in a browser

### DISCUSSION THREAD

**Discussion by amurzeau** [pradyunsg/furo#717] [Ideas]
When https://github.com/pradyunsg/furo/blob/main/src/furo/assets/scripts/furo.js is imported using a line like this in an HTML file:
`<script type="module" src="_static/scripts/furo.js"></script>`

Then Firefox will throw an error on furo.js:
`Uncaught SyntaxError: ambiguous indirect export: default`

This can be fixed by changing this line 

https://github.com/pradyunsg/furo/blob/aab86f4624b6ef38a151440be5624746d41882b1/src/furo/assets/scripts/furo.js#L1

to this:
`import * as Gumshoe from "./gumshoe-patched.js";` 

I don't know enough about javascript to tell the difference between them.
I guess `import Gumshoe from "./gumshoe-patched.js";` doesn't work with Firefox as it doesn't find any `export default` in gumshoe-patched.js, but I'm not sure about that.

So I'm asking as a feature request to change the import, if that's acceptable for you :)
The use case of this is that in Debian, as of now, furo.js is kept unmodified because not all package.json dependencies are available in Debian to transform furo.js. So packages that use Furo will get an unmodified furo.js file.

See also: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1036437

**Comment by pradyunsg** [comment]
I'm not sure how Debian is building Furo? The referenced `furo.js` isn't meant to be browser capable as-is, and going through a [build process](https://sphinx-theme-builder.readthedocs.io/en/latest/build-process/) is expected -- the sources can't be used as-is.

> I guess `import Gumshoe from "./gumshoe-patched.js";` doesn't work with Firefox

That won't work in any case, on any browser -- documentation built with Furo won't contain a `./gumshoe-patched.js` file.

> The use case of this is that in Debian, as of now, furo.js is kept unmodified because not all package.json dependencies are available in Debian to transform furo.js.

OK, in that case that's a fundamental issue in how Debian has repackaged this project? You're trying to ship an uncompiled artifact and hoping that it'd work in a browser as-is, which it won't, and you're breaking assumptions about how this package is built.

While I'm not going to volunteer to fix the Debian packaging of the project, I'm happy to take a look and point out what could be improved/changed on their end to make things work (I don't really know where to find the `rules` file for the corresponding DPKG, so I'll need a pointer!).

**Comment by amurzeau** [comment]
Thanks for your reply !

I understand, indeed it is not meant to be used like that (but it turned out to almost work "by accident").

The package is here: https://salsa.debian.org/debian/furo

But the issue is mainly missing nodejs dependencies:
- `node-autoprefixer`: available in Debian, but with an old version: Debian has 10.4.7.0 instead of >= 10.4.14
- `sass`: not available because `dart` is not available (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=998024)
  Maybe `node-sass` would be a good workaround.
- `css-loader`: available in Debian, but with an old version: Debian has 6.7.2 instead of >= 6.7.3
- `css-minimizer-webpack-plugin`: not available in Debian
- `mini-css-extract-plugin`: available but with an old version: Debian has `2.4.6` instead of >= 2.7.5
- `normalize.css`: available with a good version (8.0.1)
- `postcss-loader`: available but with an old version: Debian has 7.0.2 instead of >= 7.1.0
- `sass-loader`: not available in Debian
- `style-loader`: available but with an old version: Debian has 3.3.1 instead of >= 3.3.2
- `webpack`: available but with an old version: Debian has 5.76.1 instead of >= 5.76.3
- `webpack-cli`: available with good version 5.0.1 (bundled in `webpack` Debian package)

Many packages are just a bit older than the required version, but I guess updating them all and adding missing ones is probably a lot of work for just an offline documentation on Debian side (that happen to already mostly work).

I'm myself packaging a package for Debian that use Furo theme for its documentation (streamlink) and tried to see if it would be possible to get Furo to its full potential without too much effort, but I'm still sufficiently happy with the current state given the required work to improve it in Debian side.

Thus I'm closing this, but I'm open for discussion :)
